lib/Driver/Driver.cpp

```diff
@@ -1414,14 +1414,16 @@ void Driver::buildActions(SmallVectorImpl<const Action *> &TopLevelActions,
       }
       case types::TY_SwiftModuleFile:
       case types::TY_SwiftModuleDocFile:
-        // Module inputs are okay if generating a module.
         if (OI.ShouldGenerateModule) {
+          // When generating a .swiftmodule, treat .swiftmodule files as
+          // inputs to a MergeModule action.
           AllModuleInputs.push_back(Current);
-          break;
+        } else {
+          // Otherwise, pass .swiftmodule files as inputs to the linker,
+          // so that their debug info is available.
+          AllLinkerInputs.push_back(Current);
```

Comment:
This should probably check that we're planning to do a link action at all (`OI.shouldLink()`) and continue diagnosing otherwise.